--- desktop/CppunitTest_desktop_app.mk	2017-03-14 16:35:14.728036000 +0100
+++ desktop/CppunitTest_desktop_app.mk	2017-03-14 16:35:46.105450000 +0100
@@ -45,6 +45,12 @@
     vcl \
 ))
 
+ifeq ($(ENABLE_HEADLESS),TRUE)
+$(eval $(call gb_CppunitTest_add_libs,desktop_app,\
+	-lm $(DLOPEN_LIBS) \
+	-lpthread \
+))
+else
 ifeq ($(OS), $(filter LINUX %BSD SOLARIS, $(OS)))
 $(eval $(call gb_CppunitTest_use_static_libraries,desktop_app,\
     glxtest \
@@ -56,6 +62,7 @@
     -lX11 \
 ))
 endif
+endif
 
 $(eval $(call gb_CppunitTest_use_library_objects,desktop_app, \
     sofficeapp \
