diff -ruw ext/pdo_pgsql/pgsql_driver.stub.php ext/pdo_pgsql/pgsql_driver.stub.php
--- ext/pdo_pgsql/pgsql_driver.stub.php	2021-03-09 07:46:54.779841000 +0100
+++ ext/pdo_pgsql/pgsql_driver.stub.php	2021-03-09 07:52:52.415093000 +0100
@@ -30,4 +30,10 @@
 
     /** @return int */
     public function pgsqlGetPid() {}
+
+	/**
+     * @param callable|null $callback
+     * @return bool
+     */
+    public function pgsqlSetNoticeCallback($callback) {}
 }
