diff -ruw configure configure
--- configure
+++ configure
@@ -779,6 +779,9 @@
 enable_json1
 enable_rtree
 enable_session
+enable_pcre
+with_pcre_lib
+with_pcre_inc
 enable_debug
 enable_static_shell
 '
@@ -1445,6 +1448,7 @@
   --enable-json1          include json1 support [default=yes]
   --enable-rtree          include rtree support [default=yes]
   --enable-session        enable the session extension [default=no]
+  --enable-pcre           enable pcre handling of regexp [default=yes]
   --enable-debug          build with debugging features enabled [default=no]
   --enable-static-shell   statically link libsqlite3 into shell tool
                           [default=yes]
@@ -13326,6 +13330,107 @@
 
 
 #-----------------------------------------------------------------------
+# Do we link against libpcre and use it to implement regexp?
+TARGET_PCRE_LIBS=""
+TARGET_PCRE_INC=""
+# Check whether --enable-pcre was given.
+if test "${enable_pcre+set}" = set; then :
+  enableval=$enable_pcre; with_pcre=$enableval
+else
+  with_pcre=auto
+fi
+
+
+if test x"$with_pcre" != xno; then
+	found="yes"
+
+
+# Check whether --with-pcre-lib was given.
+if test "${with_pcre_lib+set}" = set; then :
+  withval=$with_pcre_lib; with_pcre_lib=$withval
+else
+  with_pcre_lib="auto"
+fi
+
+	if test "x$with_pcre_lib" = xauto; then
+		sLIBS="$LIBS"
+		LIBS=""
+		{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for pcre_compile in -lpcre" >&5
+$as_echo_n "checking for pcre_compile in -lpcre... " >&6; }
+if ${ac_cv_lib_pcre_pcre_compile+:} false; then :
+  $as_echo_n "(cached) " >&6
+else
+  ac_check_lib_save_LIBS=$LIBS
+LIBS="-lpcre  $LIBS"
+cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+
+/* Override any GCC internal prototype to avoid an error.
+   Use char because int might match the return type of a GCC
+   builtin and then its argument prototype would still apply.  */
+#ifdef __cplusplus
+extern "C"
+#endif
+char pcre_compile ();
+int
+main ()
+{
+return pcre_compile ();
+  ;
+  return 0;
+}
+_ACEOF
+if ac_fn_c_try_link "$LINENO"; then :
+  ac_cv_lib_pcre_pcre_compile=yes
+else
+  ac_cv_lib_pcre_pcre_compile=no
+fi
+rm -f core conftest.err conftest.$ac_objext \
+    conftest$ac_exeext conftest.$ac_ext
+LIBS=$ac_check_lib_save_LIBS
+fi
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_pcre_pcre_compile" >&5
+$as_echo "$ac_cv_lib_pcre_pcre_compile" >&6; }
+if test "x$ac_cv_lib_pcre_pcre_compile" = xyes; then :
+  TARGET_PCRE_LIBS="-lpcre"
+else
+  found="no"
+fi
+
+		LIBS="$sLIBS"
+	else
+		TARGET_PCRE_LIBS="$with_pcre_lib"
+	fi
+
+
+# Check whether --with-pcre-inc was given.
+if test "${with_pcre_inc+set}" = set; then :
+  withval=$with_pcre_inc; with_pcre_inc=$withval
+else
+  with_pcre_inc="auto"
+fi
+
+	if test "x$with_pcre_inc" = xauto; then
+		ac_fn_c_check_header_mongrel "$LINENO" "pcre.h" "ac_cv_header_pcre_h" "$ac_includes_default"
+if test "x$ac_cv_header_pcre_h" = xyes; then :
+  found="yes"
+else
+  found="no"
+fi
+
+
+	else
+		TARGET_PCRE_INC="$with_pcre_inc"
+	fi
+
+	if test x"$found" = xyes; then
+		CPPFLAGS="${CPPFLAGS} ${TARGET_PCRE_INC}"
+		LIBS="${LIBS} ${TARGET_PCRE_LIBS}"
+		BUILD_CFLAGS="${BUILD_CFLAGS} -DSQLITE_ENABLE_PCRE"
+	fi
+fi
+
+#-----------------------------------------------------------------------
 
 #-----------------------------------------------------------------------
 #   --enable-threadsafe
diff -ruw configure.ac configure.ac
--- configure.ac
+++ configure.ac
@@ -218,6 +218,23 @@
 fi
 #-----------------------------------------------------------------------
 
+#-------------------------------------------------------------------------
+#   --enable-pcre
+#
+AC_ARG_ENABLE([pcre],
+  [AC_HELP_STRING([--disable-pcre],[disable pcre even if detected])],
+  [with_pcre=$enableval],
+  [with_pcre=auto])
+AS_IF([ test x"$enable_pcre" != xno ],[
+  AC_CHECK_HEADERS([pcre.h],[
+    sLIBS=$LIBS
+    AC_SEARCH_LIBS([pcre_compile],[pcre],[
+      BUILD_CFLAGS="$BUILD_CFLAGS -DSQLITE_ENABLE_PCRE"
+    ],[])
+  ])
+])
+#-----------------------------------------------------------------------
+
 #-----------------------------------------------------------------------
 #   --enable-debug
 #
diff -ruw sqlite3.c sqlite3.c
--- sqlite3.c
+++ sqlite3.c
@@ -165133,6 +165133,9 @@
 #ifdef SQLITE_ENABLE_FTS5
 SQLITE_PRIVATE int sqlite3Fts5Init(sqlite3*);
 #endif
+#ifdef SQLITE_ENABLE_PCRE
+int sqlite3PcreInit(sqlite3 *db);
+#endif
 #ifdef SQLITE_ENABLE_JSON1
 SQLITE_PRIVATE int sqlite3Json1Init(sqlite3*);
 #endif
@@ -165169,6 +165172,10 @@
 #ifdef SQLITE_ENABLE_JSON1
   sqlite3Json1Init,
 #endif
+#ifdef SQLITE_ENABLE_PCRE
+  /* PCRE comes after ICU, so that its regexp function is preferred */
+  sqlite3PcreInit,
+#endif
 #ifdef SQLITE_ENABLE_STMTVTAB
   sqlite3StmtVtabInit,
 #endif
