--- url-0.5.9/Cargo.toml	2016-04-13 13:50:45.000000000 +0200
+++ url-0.5.9/Cargo.toml	2016-06-30 16:22:20.000000000 +0200
@@ -11,11 +11,16 @@
 keywords = ["url", "parser"]
 license = "MIT/Apache-2.0"
 
-[[test]] name = "format"
-[[test]] name = "form_urlencoded"
-[[test]] name = "idna"
-[[test]] name = "punycode"
-[[test]] name = "tests"
+[[test]]
+name = "format"
+[[test]]
+name = "form_urlencoded"
+[[test]]
+name = "idna"
+[[test]]
+name = "punycode"
+[[test]]
+name = "tests"
 [[test]]
 name = "wpt"
 harness = false
